### PR TITLE
[FEAT] Create Canvas Facade

### PR DESCRIPTION
**Pull request by Connor-Bernard**
# General Info

- [Pivotal Tracker Story](https://www.pivotaltracker.com/story/show/187434222)
- [ ] Breaking?

Note: This is a pretty massive PR, but its really just the addition of a library.  A lot of it is testing and documentation.

# Changes

- Add base facade.
- Add testing for base facade.
- Add canvas facade.
- Add testing for canvas facade.
- Add error handling for canvas facade.
- Add necessary gems for canvas facad:
    - `faraday`
    - `ostruct`
    - `timecop` (test only)

# Testing

- `spec/Facades/canvas_facade_spec.rb`
- `spec/Facades/extension_facade_base_spec.rb`

# Documentation

Doc comments have been added to all methods in the helper library.

# Checklist

- [x] Name of branch corresponds to story

### CODE REVIEW

**Comment by Connor-Bernard** [comment]
Note: codeclimate is requesting issues to fix, but they are relatively minor and will require massive amounts of refactoring.  Because of this and the quickly upcoming deadline, I think it makes more sense to ignore these issues for the time being and accrue them as tech debt to be handled later when we do not have a massive deadline.

**Review by cynthia-lixinyi — APPROVED** [review]
LGTM. As the [codeclimate](https://codeclimate.com/github/cs169/flextensions/pull/32) suggests, consider refactoring after we finish the MVP.

File: app/facades/canvas_facade.rb [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+  ##
+  # Removes a student from an existing override.
+  #
+  # @param   [Integer]    courseId the courseId to remove the student from the override of.
+  # @param   [OpenStruct] override the existing override to remove the student from.
+  # @param   [Integer]    studentId the id of the student to remove from the override.
+  # @return  [Faraday::Response] the new override if successful.
+  # @raises  [FailedPipelineError] if the student could not be removed from the override.
+  def remove_student_from_override(courseId, override, studentId)
```
Comment (ekandell) [inline app/facades/canvas_facade.rb:243]: This function never actually makes an api call, and from what I can tell neither does where it's called (line 182). If I'm understanding it right, it should call update_assignment_override so as to actually remove the student on the canvas side.

File: app/facades/canvas_facade.rb [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+      assignmentId,
+      [studentId],
+      overrideTitle,
+      newDueDate,
+      get_current_formatted_time(),
+      newDueDate
+    )
+
+    # Either successful or error that is not explicitly handled here.
+    if (createOverrideResponse.status != 400)
```
Comment (ekandell) [inline app/facades/canvas_facade.rb:148]: I'm a little confused by the error handling here, why is it only 400 codes that get handled by the rest of the function? Are we just assuming that any other type of error or a success is handled by the caller?

**Review by ekandell — COMMENTED** [review]
Thanks so much for doing this, it's a huge help! Just had a couple things that came up as I read through it

File: app/facades/canvas_facade.rb (hunk 1 above)
Comment (Connor-Bernard) [reply app/facades/canvas_facade.rb:243]: Yep!  It calls `update_assignment_override` on line 245

Comment (ekandell) [reply app/facades/canvas_facade.rb:243]: Ooops sorry lol, not sure how I missed that

File: app/facades/canvas_facade.rb [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+      raise NotFoundError.new('could not find student\'s override')
+    end
+    if (currOverride.student_ids.length == 1)
+      return update_assignment_override(
+        courseId,
+        assignmentId,
+        currOverride.id,
+        currOverride.student_ids,
+        overrideTitle,
+        get_current_formatted_time(),
```
[the remark this answers is not in the document]
Comment (Connor-Bernard) [reply app/facades/canvas_facade.rb:178]: Resolved in [6d0ff5c](https://github.com/cs169/flextensions/pull/32/commits/6d0ff5ce4cd1f9a2ea7345dc69eecba3fc04d476)

**Review by ekandell — APPROVED** [review]
LGTM!

File: config/environment.rb [hunk 4]
```diff
@@ -1,5 +1,7 @@
 # Load the Rails application.
 require_relative "application"
 
+ENV['CANVAS_URL'] ||= 'https://ucberkeley.test.instructure.com'
```
Comment (armandofox) [inline config/environment.rb:4]: don't put this in `environment.rb` as this file is not supposed to be installation-specific.  just document the fact that `CANVAS_URL` has to be set when using this adapter.

**Review by armandofox — APPROVED** [review]
this makes sense to me. i assume `ExtensionFacadeBase` is supposed to be the base/abstract class that is overridden for each type of LMS?  if so, we discussed methods that would 'link' an assignment/course/user/whatever in Flextensions to its LMS-specific counterpart - would those methods be part of this class too, or a different facade class?

**Comment by Connor-Bernard** [comment]
> this makes sense to me. i assume `ExtensionFacadeBase` is supposed to be the base/abstract class that is overridden for each type of LMS? if so, we discussed methods that would 'link' an assignment/course/user/whatever in Flextensions to its LMS-specific counterpart - would those methods be part of this class too, or a different facade class?

I think that it would be part of a different facade if I were to approach it the way I was thinking.  I like the idea of this facade being "stupid" in that it just makes API calls and does minimal error handling for them.  Likewise, it doesn't seek to understand the rest of the code-base/integration with our app.  The reason why I added these query parameters is because, on some level, every LMS is going to have a concept of these values, and they will all be necessary to provision an extension (canvas is a good example of how/why).  That said, I totally support extending the base class as well as more direct functionality from the API for all of these platforms is extended.  My goal with this PR specifically, though, was just to get the functionality in the canvas api facade set so that it could handle the entire workflow of provisioning an extension to a student (which is essentially what the function in the base class does).

File: config/environment.rb (hunk 4 above)
Comment (Connor-Bernard) [reply config/environment.rb]: Resolved in:
- [1d8ad98](https://github.com/cs169/flextensions/pull/32/commits/1d8ad9812304d313cc0eb683caa37b4934a65787)
- [979824f](https://github.com/cs169/flextensions/pull/32/commits/979824fc5dc03d630254a51b80d2200bce95c714)

**Comment by armandofox** [comment]
re "not really handling errors": one thing the facade does need to do is rescue any errors on the remote side and wrap them in the generic facade's error type(s).  looks like you have a start on this already.

**Comment by Connor-Bernard** [comment]
> re "not really handling errors": one thing the facade does need to do is rescue any errors on the remote side and wrap them in the generic facade's error type(s). looks like you have a start on this already.

Oh yeah, I rescue all rails errors thrown.  I just don't have dedicated error handling for if the http response is an error.  Most of these functions in the facade just return the response itself, so I was leaving it up to the next facade / class that uses these to handle it how they see fit for their specific use case.

**Review by ekandell — APPROVED** [review]
LGTM

**Pull request merged**